src/routes/article-list.ts

```diff
@@ -1,28 +1,49 @@
 import Router from '@koa/router';
 import { Context, Middleware, Next } from 'koa';
-import { hydra, rdf, schema } from 'rdf-namespaces';
-import Articles from '../articles';
+import { Store as N3Store } from 'n3';
+import { DataFactory, Quad, Store } from 'rdf-js';
+import url from 'url';
+import { hydra, rdf, schema } from '../namespaces';
+import streamToArray from '../stream-to-array';
 import Routes from './index';
 
-export default (articles: Articles, router: Router): Middleware => (
+export default (
+  articles: Store,
+  router: Router,
+  {
+    blankNode, namedNode, literal, quad,
+  }: DataFactory,
+): Middleware => (
   async ({ request, response }: Context, next: Next): Promise<void> => {
-    response.body = {
-      '@context': {
-        '@base': request.origin,
-      },
-      '@id': router.url(Routes.ArticleList),
-      '@type': hydra.Collection,
-      [hydra.title]: { '@value': 'List of articles', '@language': 'en' },
-      'http://www.w3.org/ns/hydra/core#manages': {
-        'http://www.w3.org/ns/hydra/core#property': { '@id': rdf.type },
-        'http://www.w3.org/ns/hydra/core#object': { '@id': schema.Article },
-      },
-      [hydra.totalItems]: await articles.count(),
-      [hydra.member]: {
-        '@list': [...articles],
-      },
-    };
-    response.type = 'jsonld';
+    const articleList = namedNode(url.resolve(request.origin, router.url(Routes.ArticleList)));
+
+    const manages = blankNode('manages');
+
+    const quads = [
+      quad(articleList, rdf('type'), hydra('Collection')),
+      quad(articleList, hydra('title'), literal('List of articles', 'en')),
+      quad(articleList, hydra('manages'), manages),
+      quad(manages, hydra('property'), rdf('type')),
+      quad(manages, hydra('object'), schema('Article')),
+    ];
+    const found = await streamToArray(articles.match(null, rdf('type'), schema('Article')));
+
+    const newQuads = found.reduce(
+      async (carryPromise: Promise<Array<Quad>>, { subject }: Quad): Promise<Array<Quad>> => {
+        const carry = await carryPromise;
+
+        carry.push(quad(articleList, hydra('member'), subject));
+
+        (await streamToArray(articles.match(null, null, null, subject)))
+          .forEach(({ predicate, object }: Quad): void => {
+            carry.push(quad(subject, predicate, object));
+          });
+
+        return carry;
+      }, Promise.resolve(quads),
+    );
```

Comment:
This is very messy. Think can use regex for the second match to reduce it to one call, but don't think it's possible to directly paginate on the first.

Reply:
Thinking we'll still want an `Articles`, which could return a `Stream` or a `Dataset`.